File: scripts/build_aaz_index.py
#!/usr/bin/env python3
"""Build standalone aaz-index.html from template + app JS + JSON module manifest."""

from __future__ import annotations

import argparse
import base64
import hashlib
import json
from pathlib import Path
from typing import Any

from app_data_modules import assemble_from_modules, canonical_json_text


def compute_build_id(data_str: str, js: str, tpl: str) -> str:
    def normalize(value: str) -> str:
        return value.replace("\r\n", "\n").replace("\r", "\n")

    digest = hashlib.sha1()
    digest.update(normalize(data_str).encode("utf-8"))
    digest.update(b"\0")
    digest.update(normalize(js).encode("utf-8"))
    digest.update(b"\0")
    digest.update(normalize(tpl).encode("utf-8"))
    return digest.hexdigest()[:12]


def escape_json_for_html_script(json_text: str) -> str:
    """Escape JSON for safe embedding inside <script type="application/json">."""
    return (
        json_text
        .replace("</script", "<\\/script")
        .replace("<!--", "<\\!--")
    )


def app_data_module_manifest_text(modules_dir: Path, build_id: str) -> str:
    manifest_path = modules_dir / "manifest.json"
    manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
    if not isinstance(manifest, dict):
        raise ValueError(f"JSON root must be object: {manifest_path}")
    modules = manifest.get("modules", [])
    if not isinstance(modules, list) or not modules:
        raise ValueError(f"App data module manifest is empty: {manifest_path}")

    enriched: list[dict[str, Any]] = []
    for item in modules:
        if not isinstance(item, dict):
            continue
        file_name = str(item.get("file", "")).strip()
        if not file_name:
            raise ValueError("App data module entry is missing file")
        text = (modules_dir / file_name).read_text(encoding="utf-8").replace("\r\n", "\n").replace("\r", "\n")
        raw = text.encode("utf-8")
        enriched.append(
            {
                "file": file_name,
                "keys": item.get("keys") if isinstance(item.get("keys"), list) else [],
                "bytes": len(raw),
                "sha256": base64.b64encode(hashlib.sha256(raw).digest()).decode("ascii"),
            }
        )

    payload = {
        "mode": "modules",
        "version": manifest.get("version") or 1,
        "build_id": build_id,
        "base_url": "./data/modules/",
        "modules": enriched,
        "key_order": manifest.get("key_order") if isinstance(manifest.get("key_order"), list) else [],
    }
    return json.dumps(payload, ensure_ascii=False, indent=2) + "\n"


def build(
    data_path: Path,
    js_path: Path,
    template_path: Path,
    out_path: Path,
    build_id: str | None = None,
    modules_dir: Path | None = None,
) -> str:
    if data_path.exists():
        data_obj = json.loads(data_path.read_text(encoding="utf-8"))
        if not isinstance(data_obj, dict):
            raise ValueError(f"JSON root must be object: {data_path}")
        data_str = canonical_json_text(data_obj)
    elif modules_dir is not None:
        data_obj = assemble_from_modules(modules_dir)
        data_str = canonical_json_text(data_obj)
    else:
        raise FileNotFoundError(f"Data file not found: {data_path}")
    js = js_path.read_text(encoding="utf-8")
    tpl = template_path.read_text(encoding="utf-8")
    resolved_build_id = (build_id or "").strip() or compute_build_id(data_str, js, tpl)
    js = js.replace("__APP_BUILD_ID__", resolved_build_id)
    app_data_payload = (
        app_data_module_manifest_text(modules_dir, resolved_build_id)
        if modules_dir is not None
        else data_str
    )
    html = (
        tpl
        .replace("__APP_DATA_JSON__", escape_json_for_html_script(app_data_payload))
        .replace("__APP_SCRIPT__", js)
    )
    out_path.write_text(html, encoding="utf-8-sig")
    return resolved_build_id


def main() -> int:
    parser = argparse.ArgumentParser(description="Build standalone aaz-index.html")
    parser.add_argument("--data", default="app_data.json", help="Path to app_data.json")
    parser.add_argument("--modules-dir", default="data/modules", help="Directory with split app_data modules")
    parser.add_argument("--js", default="v3_app.js", help="Path to v3_app.js")
    parser.add_argument("--template", default="v3_template.html", help="Path to v3_template.html")
    parser.add_argument("--out", default="aaz-index.html", help="Output HTML path")
    parser.add_argument("--build-id", default="", help="Optional explicit build id (default: deterministic hash of data+js+template)")
    args = parser.parse_args()

    build_id = build(
        data_path=Path(args.data),
        js_path=Path(args.js),
        template_path=Path(args.template),
        out_path=Path(args.out),
        build_id=args.build_id,
        modules_dir=Path(args.modules_dir) if str(args.modules_dir).strip() else None,
    )
    print(f"OK: built {args.out} (build_id={build_id})")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/build_aaz_index.mjs
import { createHash } from 'node:crypto';
import { readFileSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';

function parseArgs(argv) {
  const args = {
    data: 'app_data.json',
    js: 'v3_app.js',
    template: 'v3_template.html',
    out: 'aaz-index.html',
    buildId: '',
  };
  for (let i = 0; i < argv.length; i += 1) {
    const key = argv[i];
    const value = argv[i + 1] || '';
    if (key === '--data') args.data = value;
    if (key === '--js') args.js = value;
    if (key === '--template') args.template = value;
    if (key === '--out') args.out = value;
    if (key === '--build-id') args.buildId = value;
    if (key.startsWith('--')) i += 1;
  }
  return args;
}

function canonicalJsonText(dataPath) {
  const data = JSON.parse(readFileSync(dataPath, 'utf8'));
  if (!data || Array.isArray(data) || typeof data !== 'object') {
    throw new Error(`JSON root must be object: ${dataPath}`);
  }
  return `${JSON.stringify(data, null, 2)}\n`;
}

function computeBuildId(dataText, jsText, templateText) {
  const normalizeLineEndings = (value) => String(value || '').replace(/\r\n?/g, '\n');
  return createHash('sha1')
    .update(normalizeLineEndings(dataText), 'utf8')
    .update('\0', 'utf8')
    .update(normalizeLineEndings(jsText), 'utf8')
    .update('\0', 'utf8')
    .update(normalizeLineEndings(templateText), 'utf8')
    .digest('hex')
    .slice(0, 12);
}

function escapeJsonForHtmlScript(jsonText) {
  return jsonText
    .replaceAll('</script', '<\\/script')
    .replaceAll('<!--', '<\\!--');
}

function cspSha256(text) {
  const browserScriptText = String(text || '').replace(/\r\n?/g, '\n');
  return `'sha256-${createHash('sha256').update(browserScriptText, 'utf8').digest('base64')}'`;
}

function inlineScriptHashes(html) {
  const hashes = [];
  const scriptRe = /<script\b(?![^>]*\bsrc=)[^>]*>([\s\S]*?)<\/script>/gi;
  for (const match of html.matchAll(scriptRe)) {
    hashes.push(cspSha256(match[1]));
  }
  return hashes;
}

const args = parseArgs(process.argv.slice(2));
const dataText = canonicalJsonText(args.data);
const templateText = readFileSync(args.template, 'utf8');
let jsText = readFileSync(args.js, 'utf8');
const buildId = args.buildId.trim() || computeBuildId(dataText, jsText, templateText);
jsText = jsText.replaceAll('__APP_BUILD_ID__', buildId);
let html = templateText
  .split('__APP_DATA_JSON__').join(escapeJsonForHtmlScript(dataText))
  .split('__APP_SCRIPT__').join(jsText);
const scriptHashes = inlineScriptHashes(html);
html = html.split('__CSP_SCRIPT_HASHES__').join(scriptHashes.join(' '));

writeFileSync(args.out, `\uFEFF${html}`, 'utf8');
console.log(`OK: built ${join(process.cwd(), args.out)} (build_id=${buildId})`);
